Solves ctrl+click on functions by ignoring the cursor

2shady4u opened the pull request:
Solves #31036, #20393 and issues discussed in #7080 (partially)

I went through the dismal spaghetti and conquered the legacy.
This PR solves an issue with blocking behavior of TK_CURSOR when control+clicking on an identifier (to jump to the function definition).

Solution at work:
(This is not possible with the current gdscript editor!)
![fix](https://user-images.githubusercontent.com/42484461/62482043-5f288c80-b7b4-11e9-9c01-3dcce4d84b7f.gif)

**What does it solve?**
Control+clicking on a function call should normally jump to wherever the function definition is placed. This jump did not happen when the function definition was placed after the function call (see #31036). 

**What was the problem?**
The parser expects the next token to be an open parenthesis and the current one to be a literal for proper parsing purposes. Now when the cursor is also part of the token array, this cursor blocks and exits the parser. The result is that ONLY the functions that were defined BEFORE the cursor would be added to the functions array of the Parser. As a result... finding functions that were defined AFTER the cursor was impossible.

**Is this the best implementation?**
I think there might be a more elegant solution somewhere 🤔 (or not)
Please guide me into the right direction!

Cheerios!

akien-mga commented:
Note that you can set up `clang-format` locally to avoid ping pong with Travis checks :)
http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html

2shady4u commented:
So I wanted to clarify some things regarding the implementation of this PR:
- Errors of the parser aren't printed to the console so the user has no idea on what's actually going on inside of the parser. This is most likely deliberate so that the user doesn't get overwhelmed by the sheer amount of "false flag" parser errors. 
This part of the PR is actually temporary as it doesn't make sense for this particular error to be printed while the rest stays hidden:
````	
if (error_set) {
			print_line(error);
			return;
		}`
```
Nevertheless I wanted to know if there's a parser log file somewhere in which these errors get dumped or if there isn't: why not?

- The issue that this PR tries to address is mostly related to the fact that the tokenizer gives back the TK_CURSOR while in this part of the code, the cursor should actually always be ignored.
Might be easier to have:
- A boolean inside of the tokenizer that says something like "ignore_cursor" and thus doesn't ever return TK_CURSOR when this boolean is high.
- An option to give an optional boolean to the get_token function, in the following form: ```tokenizer->get_token(p_offset can_be_cursor = true)```
- The current implementation (this PR) which doesn't break compatibility and is very clear, but bloats the code.

I don't really know Godot's mantra regarding this? @bojidar-bg @vnen

2shady4u commented:
Cleaned it up a little bit.

Calinou commented:
> Nevertheless I wanted to know if there's a parser log file somewhere in which these errors get dumped or if there isn't: why not?

Godot doesn't log standard output/error to files by default, but this can be enabled in the **File Logging** section of the Project Settings. Still, for parser errors to be logged, they need to be outputted to standard output (or standard error) in the first place. Maybe they could be added as `print_verbose()` calls, so that they would only be printed when starting the editor with the `--verbose` command-line argument.

KoBeWi commented:
Sooo, I'm using a custom Godot build with few yet-unmerged PRs (including this one) and sometimes the editor hangs and spams this error into the console:
```
ERROR: GDScriptTokenizerText::get_token: Condition ' p_offset >= MAX_LOOKAHEAD ' is true. returned: TK_ERROR
   At: modules\gdscript\gdscript_tokenizer.cpp:1122
```
Any chance this could be caused by your PR? I might try to isolate the issue if it's related.

2shady4u commented:
Hello KoBeWi,
It probably is due to the while loop I implemented.
However I do need this while-loop to make it work because the
next literal (which is supposedly found after a few iterations) needs to be found.

Seems that in your case the while-loop has difficulty finding the next literal and
goes out of bounds...

Do you have a .gd-script that demonstrates this issue? So I can try fixing it.

Also, I'm still waiting for someone with more experience to actually say that the way I implemented this is either bad or good. So the actual implementation might totally change in the future.

KoBeWi commented:
I managed to extract this piece of code:
```
func test(map, loaded_maps, deep):
	for neighbor in map.map_data.neighbors:
		if neighbor in loaded_maps:
			if deep:
				print(loaded_maps[neighbor])
```
Put the cursor after the first if and write `nei` (so the line looks like `if neineighbor in loaded_maps:`). The editor will hang.

2shady4u commented:
Ok, I removed the while-loop as it was substantial overkill.
KoBeWi's example code works now.

There are actually only 3 cases:
1. The next token is not the Cursor --> next_valid_offset should be 1
2. The next token is the Cursor, again 2 options appear:
2a. The Cursor is at the end of the Identifier and the next-next token is an open Parethesis (
--> next_valid_offset should be 2
2a. The Cursor is somewhere in the middle of the Identifier and the next-next token is a chunk of the identifier.
--> next_valid_offset should be 3

Cheerios!
